merkle-tree/src/lib.rs

```diff
@@ -35,13 +33,23 @@ impl<L, D> MerkleTree<L, D> {
     pub fn new<H, C>(h: &H, c: &C, leaves: Vec<RowMajorMatrix<L>>) -> Self
     where
         L: Copy,
-        D: Copy + Default,
+        D: Copy + Default + PartialEq,
         H: CryptographicHasher<L, D>,
         C: PseudoCompressionFunction<D, 2>,
     {
         assert!(!leaves.is_empty(), "No matrices given?");
 
-        // TODO: Check the matching height condition.
+        // check height property
+        assert!(
+            leaves
+                .iter()
+                .map(|m| m.height())
+                .sorted_by_key(|&h| Reverse(h))
```

Comment:
nit: could this be `sorted()` with no key? Seems like reversing should be optional